gc/base/standard/ConcurrentGCSATB.cpp

```diff
@@ -554,4 +554,36 @@ MM_ConcurrentGCSATB::reportConcurrentCollectionStart(MM_EnvironmentBase *env)
 	}
 }
 
+/**
+ * TLH is about to be cleared, we must premark it if SATB is active. This is
+ * non-trival to do, we must know the start of the last object to mark the proper range of the TLH.
+ * Arbitrary bits at the end of the TLH can’t be marked, this is problematic for sweep.
+ *
+ * To get around this issue we seal the TLH with a bogus/dummy obj at the
+ * end of the TLH. With this, we know the precise addr to batch mark to.
+ *
+ * We have guaranteed that we have at least min obj size available in the TLH by reservering
+ * it when the TLH was initially allocated.
+ */
+void
+MM_ConcurrentGCSATB::allocCacheRestored(MM_EnvironmentBase *env, MM_TLHAllocationSupport *tlh) {
+	if (_extensions->isSATBBarrierActive()) {
+		/* Mark all newly allocated objects */
+		if (tlh->containsObjs()) {
+			/* We are guaranteed to have the required size, this must have been reserved during TLH refresh. */
+			Assert_MM_true(tlh->getRemainingSize() >= reservedForGCAllocCacheSize());
+
+			void* lastTLHobj = tlh->getAlloc();
+
+			_extensions->objectModel.initializeMinimumSizeObject(env, lastTLHobj);
+
+			tlh->setAlloc((void *)((uintptr_t) lastTLHobj + OMR_MINIMUM_OBJECT_SIZE));
```

Comment:
we could do one more level of indirection here and do something like

if (satbActive)
if (tlh->liveObjectBoundaries(&base, &top)) 
markObjectsForRange(base, top)

instead of exposing internals and a bunch of private TLH API's

Reply:
although, it would be hard to keep remainingSize assert :(

Reply:
But since reservedForGCAllocCacheSize relies on runtime check in Ext if SATB is enabled, there is no need to keep this method in GlobalCollector - the virtual dispatch (to filter our IncrementalUpdate variant) is a complete waste of time.

Reply:
> But since reservedForGCAllocCacheSize relies on runtime check in Ext if SATB is enabled, there is no need to keep this method in GlobalCollector - the virtual dispatch (to filter our IncrementalUpdate variant) is a complete waste of time.

Considering the following method that uses `reservedForGCAllocCacheSize()`: 

```C++
void
MM_TLHAllocationSupport::reserveTLHTopForGC(MM_EnvironmentBase *env)
{
	uintptr_t bytesToReserveForGC = env->getExtensions()->getGlobalCollector()->reservedForGCAllocCacheSize();

	Assert_MM_true(_reservedBytesForGC == 0);

	if (bytesToReserveForGC > 0) {
		_reservedBytesForGC = bytesToReserveForGC;
		setTop((void *) ((uintptr_t) getTop() - _reservedBytesForGC));
	}
}
```

Are you suggesting we change line 
```Diff
- uintptr_t bytesToReserveForGC = env->getExtensions()->getGlobalCollector()->reservedForGCAllocCacheSize();
+ uintptr_t bytesToReserveForGC = (_extensions->isSATBBarrierActive() ? OMR_MINIMUM_OBJECT_SIZE : 0);
```

we'd also lose this assert:
```
MM_TLHAllocationSupport::allocateFromTLH(...)
...
Assert_MM_true(_reservedBytesForGC == extensions->getGlobalCollector()->reservedForGCAllocCacheSize());
```


The other option is actually move `reservedForGCAllocCacheSize()` to the base class with isSATBactive check then it can be removed it from SATB base collector.